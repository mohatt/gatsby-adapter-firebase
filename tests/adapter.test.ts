import { createGatsbyReporter, mountTestProject, projectRoot } from '../test/util.js'
import createAdapter from '../src/index.js'

describe('adapter()', () => {
  let cwdSpy: ReturnType<typeof vi.spyOn>

  beforeEach(() => {
    cwdSpy = vi.spyOn(process, 'cwd').mockReturnValue(projectRoot)
  })

  afterEach(() => {
    cwdSpy.mockRestore()
    vi.unstubAllEnvs()
  })

  it('builds test-project with default options', async () => {
    const { vol, adaptArgs } = await mountTestProject('test-project')
    const adapter = createAdapter()
    expect(adapter).toHaveProperty('name', 'gatsby-adapter-firebase')

    const reporter = createGatsbyReporter()
    await expect(adapter.config({ reporter } as any)).resolves.toMatchSnapshot('config')

    await expect(adapter.adapt({ ...adaptArgs, reporter } as any)).resolves.toBe(undefined)

    await expect(vol).toMatchVolumeSnapshot('test-project', {
      prefix: projectRoot,
      report: 'all',
    })

    // assert reporter calls
    expect(reporter.verbose.mock.calls).toMatchSnapshot('reporter.verbose')
    expect(reporter.activityTimer.mock.calls).toMatchSnapshot('reporter.activityTimer')
    expect(reporter.activity.setStatus.mock.calls).toMatchSnapshot(
      'reporter.activityTimer.setStatus',
    )
  })

  it('builds test-project with custom options', async () => {
    const { vol, adaptArgs } = await mountTestProject('test-project')
    const adapter = createAdapter({
      hostingTarget: 'custom-hosting',
      functionsOutDir: 'functions/custom',
      functionsCodebase: 'custom-functions',
      functionsRuntime: 'nodejs22',
      functionsConfig: {
        labels: {
          custom: 'custom-label',
        },
      },
      functionsConfigOverride: {
        'ssr-engine': {
          timeoutSeconds: 120,
          region: 'asia-northeast1',
        },
      },
      storageBucket: 'test-project.appspot.com',
      excludeDatastoreFromEngineFunction: true,
    })

    vi.stubEnv('DEPLOY_URL', 'https://test.local')
    const reporter = createGatsbyReporter()
    await expect(adapter.config({ reporter } as any)).resolves.toMatchSnapshot('config')

    await expect(adapter.adapt({ ...adaptArgs, reporter } as any)).resolves.toBe(undefined)

    await expect(vol).toMatchVolumeSnapshot('test-project-custom', {
      prefix: projectRoot,
      report: 'all',
    })
  })
})
