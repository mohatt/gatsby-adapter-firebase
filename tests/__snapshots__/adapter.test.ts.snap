// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`adapter() > builds test-project with custom options > config 1`] = `
{
  "deployURL": "https://test.local",
  "excludeDatastoreFromEngineFunction": true,
  "functionsArch": "x64",
  "functionsPlatform": "linux",
  "pluginsToDisable": [
    "gatsby-plugin-netlify-cache",
    "gatsby-plugin-netlify",
  ],
  "supports": {
    "pathPrefix": true,
    "trailingSlash": [
      "always",
      "never",
      "ignore",
    ],
  },
}
`;

exports[`adapter() > builds test-project with default options > config 1`] = `
{
  "deployURL": undefined,
  "excludeDatastoreFromEngineFunction": false,
  "functionsArch": "x64",
  "functionsPlatform": "linux",
  "pluginsToDisable": [
    "gatsby-plugin-netlify-cache",
    "gatsby-plugin-netlify",
  ],
  "supports": {
    "pathPrefix": true,
    "trailingSlash": [
      "always",
      "never",
      "ignore",
    ],
  },
}
`;

exports[`adapter() > builds test-project with default options > reporter.activityTimer 1`] = `
[
  [
    "[gatsby-adaptor-firebase]: Building functions workspace",
  ],
  [
    "[gatsby-adaptor-firebase]: Building hosting config",
  ],
  [
    "[gatsby-adaptor-firebase]: Building firebase.json",
  ],
]
`;

exports[`adapter() > builds test-project with default options > reporter.activityTimer.setStatus 1`] = `
[
  [
    "codebase=gatsby, files=26, functions=5 (use --verbose for breakdown)",
  ],
  [
    "target=gatsby, redirects=2, rewrites=7, headers=9",
  ],
  [
    "updated (5.13 KB)",
  ],
]
`;

exports[`adapter() > builds test-project with default options > reporter.verbose 1`] = `
[
  [
    "[gatsby-adaptor-firebase]: version: 0.0.0-test",
  ],
]
`;
